pennylane/ops/op_math/prod.py

```diff
@@ -415,6 +427,162 @@ def _swappable_ops(op1, op2, wire_map: dict = None) -> bool:
     if wire_map is not None:
         wires1 = wires1.map(wire_map)
         wires2 = wires2.map(wire_map)
-    if np.intersect1d(wires1, wires2).size != 0:
-        return False
-    return np.min(wires1) > np.min(wires2)
+    wires1 = set(wires1)
+    wires2 = set(wires2)
+    return False if wires1 & wires2 else wires1.pop() > wires2.pop()
+
+
+class ProductFactorsGrouping:
+    """Utils class used for grouping identical product factors."""
+
+    _pauli_mult = {
+        "II": (1.0, "I"),
+        "IX": (1.0, "X"),
+        "IY": (1.0, "Y"),
+        "IZ": (1.0, "Z"),
+        "XI": (1.0, "X"),
+        "XX": (1.0, "I"),
+        "XY": (1.0j, "Z"),
+        "XZ": (-1.0j, "Y"),
+        "YI": (1.0, "Y"),
+        "YX": (-1.0j, "Z"),
+        "YY": (1.0, "I"),
+        "YZ": (1.0j, "X"),
+        "ZI": (1.0, "Z"),
+        "ZX": (1.0j, "Y"),
+        "ZY": (-1.0j, "X"),
+        "ZZ": (1.0, "I"),
+    }
+    _paulis = {"X": PauliX, "Y": PauliY, "Z": PauliZ}
+
+    def __init__(self):
+        self._pauli_factors = {}  #  {wire: (pauli_coeff, pauli_word)}
+        self._non_pauli_factors = {}  # {wires: [hash, pow_coeff, operator]}
+        self.factors = ()
+        self.global_phase = 1
+
+    def add(self, factor: Operator):
+        """Add factor.
+
+        Args:
+            factor (Operator): Factor to add.
+        """
+        if isinstance(factor, Prod):
+            for prod_factor in factor.factors:
+                self.add(prod_factor)
+        elif isinstance(factor, Sum):
+            self.factors += (factor.summands,)
+        elif not isinstance(factor, qml.Identity):
+            if isinstance(factor, SProd):
+                self.global_phase *= factor.scalar
+                factor = factor.base
+            wires = factor.wires
+            if isinstance(factor, (qml.Identity, qml.PauliX, qml.PauliY, qml.PauliZ)):
+                self._add_pauli_factor(factor=factor, wires=wires)
+                self._remove_non_pauli_factors(wires=wires)
+            else:
+                self._add_non_pauli_factor(factor=factor, wires=wires)
+                self._remove_pauli_factors(wires=wires)
+
+    def _add_pauli_factor(self, factor: Operator, wires: List[int]):
+        """Adds the given Pauli operator to the temporary ``self._pauli_factors`` dictionary. If
+        there was another Pauli operator acting on the same wire, the two operators are grouped
+        together using the ``self._pauli_mult`` dictionary.
+
+        Args:
+            factor (Operator): Factor to be added.
+            wires (List[int]): Factor wires. This argument is added to avoid calling
+                ``factor.wires`` several times.
+        """
+        wire = wires[0]
+        label = factor.label()
```

Comment:
Probably better to use `name` or `base_name` in this case.   Even maybe `type(op).__name__`.

Reply:
Why? `label()` is as fast as `name`. Also since `label()` returns `X` instead of `PauliX`, we need less strings to identify the classes.

Reply:
You are right, I will change it to `name` then.

Reply:
Done!